### DISCUSSION TITLE
[Compilers] Source code for path `...` not found

### DISCUSSION THREAD

**Discussion by ulerdogan** [zkSync-Community-Hub/zksync-developers#314] [Compilers] [answered]
### Which compiler are you using?

- [X] zksolc
- [ ] zkvyper

### Which version of Solidity or Vyper?

0.8.17

### zksolc/zkvyper version?

v1.3.22

### Explain the issue

Hello,

I am having an issue similar to the one in [discussion#208](https://github.com/zkSync-Community-Hub/zksync-developers/discussions/208), opening a new discussion thread as recommended by @dutterbutter. My goal is compiling [our repo](https://github.com/getclave/clave-contracts) with `zksolc`, i am running compilers in a container.

My steps to reproduce:

In terminal

 ``` bash
SOLC_VERSION="0.8.17-alpine"

docker create -it --name zksolc --entrypoint ash  ethereum/solc:${SOLC_VERSION}

docker cp clave-contracts zksolc:/clave-contracts

docker start -i zksolc
```

In the container

``` bash
ZKSOLC_VERSION="v1.3.22"

wget https://github.com/matter-labs/zksolc-bin/raw/main/linux-amd64/zksolc-linux-amd64-musl-${ZKSOLC_VERSION} -O /bin/zksolc; chmod +x /bin/zksolc

zksolc -O3 \
    --base-path clave-contracts \
    --include-path clave-contracts/node_modules \
    --combined-json abi,bin \
    -o clave-contracts/build \
    clave-contracts/contracts/**/*.sol
```

Then, having this issue 

``` bash
...

Source code for path `@matterlabs/zksync-contracts/l2/system-contracts/BootloaderUtilities.sol` not found
```

Additionally, if I run `zksolc` in my MACOS device, I receive the same error with a different file path:

``` bash
...

Source code for path `@account-abstraction/contracts/samples/callback/TokenCallbackHandler.sol` not found
```

### Link to your repo?

[https://github.com/getclave/clave-contracts](https://github.com/getclave/clave-contracts)

**Comment by dutterbutter** [comment]
Thanks @ulerdogan for opening this in a separate discussion! 

I was able to compile the contracts from the repo https://github.com/getclave/clave-contracts fine locally.

To reproduce:
- Yarn install
- Yarn hardhat compile

 It is likely the issue is specific to the Docker setup. It is likely the dependencies are not being copied / installed in the container. I see the reference to them in the start command but the issue is likely there. Can you ssh into the container and see the dependencies that are being reported as missing?


 Is there a particular reason for this unique compiler in container setup?

**Comment by ulerdogan** [comment]
Yes, it's being compiled with Hardhat, but my goal is generating bindings with `abigen`. So, i need `combined-json` files directly from `solc`/`zksolc`. I don't have a Dockerfile setup because just using for utilizing more specific versions separated from my OS. Having the same issue without Docker too by directly installing `zksolc`. I checked both in container or not, dependencies are complete.

For the versions, it's the latest version of `zksolc` and the  Solidity version we use in mainnet. Also, followed the same steps for the EVM version of the repo by compiling with `solc` instead of `zksolc`. There is no issue.

Thanks!

**Reply by dutterbutter** [reply]
I see. I was able to reproduce on my side. Will continue to investigate.

**Reply by ulerdogan** [reply]
Thanks!

**Reply by hedgar2017** [reply, marked as answer]
Hey @ulerdogan, the issue is that `zksolc` doesn't operate on the Solidity source code level so it doesn't resolve dependencies passed with `--include-path` nor it includes them into the list of files for compilation.
Could you try passing required dependencies from `node_modules` explictly after `clave-contracts/contracts/**/*.sol`?

**Comment by lysumeduso** [comment]
It seems that the zksolc compiler is unable to locate the source code for the specified paths. Make sure that the paths to the source files are correct and accessible from within the container. Additionally, ensure that the specified base path and include paths are correctly configured to point to the source files and dependencies. Double-check the file paths and try running the zksolc command again.

**Comment by tus4ncockverb** [comment]
Hello! The problem you are facing is that the zkSync compiler cannot find the source code for some files specified in your project. This may be caused by an incorrect directory structure or incorrect file paths. Make sure that the file paths are correct and that all required dependencies are included, including @matterlabs/zksync-contracts и @account-abstraction, installed and available to the compiler.

**Comment by quanteauphil** [comment]
It seems like the issue is related to the paths used to import Solidity files in your contracts. The compiler is unable to locate the source code for certain paths, such as @matterlabs/zksync-contracts/l2/system-contracts/BootloaderUtilities.sol and @account-abstraction/contracts/samples/callback/TokenCallbackHandler.sol.

Make sure that the paths specified in your import statements are correct and that the necessary files are available in the specified locations. You may need to adjust the base path and include path parameters provided to zksolc to ensure that the compiler can find the required Solidity files.

Double-check your project's directory structure and the paths used in your import statements to ensure consistency and correctness.

**Comment by hedgar2017** [comment]
@lysumeduso @tus4ncockverb @quanteauphil please avoid posting AI-generated comments in zkSync-Community-Hub.
Repeated actions will result in limiting access to these resources.
Thank you for cooperation!

**Comment by mcgeesusan** [comment]
It seems that zksolc is unable to locate the source code for some imported contracts. Ensure that the paths specified in your compiler command match the actual directory structure in your repository. Double-check the paths for @matterlabs/zksync-contracts and @account-abstraction/contracts to ensure they are correctly mapped within your project.

**Comment by matthewredmanv** [comment]
It seems that zksolc is unable to find the source code files specified in your contract imports. Make sure that the paths to the imported files are correct and accessible from the location where zksolc is being run. Double-check the base path and include path configurations to ensure they point to the correct directories containing the source files.

**Comment by queskim** [comment]
The issue you are facing is that zksolc is unable to find the source code for the specified paths. You may need to check your file paths and ensure they are correct. Additionally, you can try using the --allow-paths flag to explicitly allow zksolc to access the specified paths.
